### PR TITLE
bluetooth: fast_pair: fp_storage: Change overwriting Account Key scheme

### PR DESCRIPTION
**Pull request by alstrzebonski**
Earlier, the oldest Account Key was being overwrited by new one. Now, we overwrite the least recently used Account Key as according to the Fast Pair specification. New Settings entry has been introduced to keep track of which Key to overwrite.

TODO: Needs edge case testing and tests alignment.

Jira: NCSDK-21061

### CODE REVIEW

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 1]
```diff
@@ -309,22 +404,30 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
...
 	if (account_key_count < ACCOUNT_KEY_CNT) {
 		account_key_count++;
 	}
 
+	ak_order_update(id);
+	err = settings_save_one(SETTINGS_AK_ORDER_FULL_NAME, account_key_order,
+				sizeof(account_key_order));
+	if (err) {
+		LOG_ERR("Unable to save new Account Key order in Settings."
+			"Not propagating the error and keeping correct Account Key order in RAM."
```
Comment (MarekPieta) [inline subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:416]: `keeping updated Account Key order in RAM`

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 2]
```diff
@@ -118,31 +141,105 @@ static bool zero_id_check(uint8_t start_idx)
...
+		if (account_key_loaded_ids[i] != ACCOUNT_KEY_MIN_ID + i) {
 			return false;
 		}
 	}
 
 	return true;
 }
 
-static bool rollover_check(uint8_t cur_id, uint8_t prev_id)
+static uint8_t toggle_ak_id(uint8_t id)
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:155]: I would rather keep it as it is. We toggle ID when we substitute ID (e.g. 2 -> 7 -> 2 -> 7 -> ...).

[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:155]: Maybe `update_ak_id`/`bump_ak_id`?

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 3]
```diff
@@ -118,31 +141,105 @@ static bool zero_id_check(uint8_t start_idx)
...
+		bool id_found = false;
+
+		for (size_t j = 0; j < account_key_count; j++) {
+			if (account_key_order[j] == id) {
+				id_found = true;
+				break;
+			}
+		}
+		if (!id_found) {
+			__ASSERT_NO_MSG(ak_order_update_count < account_key_count);
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:219]: In theory, the assert is impossible to trigger and it can be opt out in release builds, but maybe it will be safer to add return here as it don't cost much.

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 4]
```diff
@@ -118,31 +141,105 @@ static bool zero_id_check(uint8_t start_idx)
...
+			ak_order_update(id);
+			ak_order_update_count++;
+			/* Start loop again to check whether after update no existent AK has been
+			 * lost.
+			 */
+			i = 0;
+		}
+	}
+
+	err = settings_save_one(SETTINGS_AK_ORDER_FULL_NAME, account_key_order,
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:229]: My initial thought was to save empty list on first boot, but it actually is not needed.

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:155]: bump_ak_id sounds good

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 5]
```diff
@@ -247,6 +324,22 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
...
 		if (account_key_check_cb(&account_key_list[i], context)) {
+			uint8_t old_account_key_order[sizeof(account_key_order)];
+			int err;
+
+			memcpy(old_account_key_order, account_key_order, sizeof(account_key_order));
+			ak_order_update(account_key_loaded_ids[i]);
+			err = settings_save_one(SETTINGS_AK_ORDER_FULL_NAME, account_key_order,
+						sizeof(account_key_order));
+			if (err) {
+				LOG_ERR("Unable to save new Account Key order in Settings."
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:335]: That seems like a good idea.

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 6]
```diff
@@ -309,22 +404,30 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
...
 		account_key_count++;
 	}
 
+	ak_order_update(id);
+	err = settings_save_one(SETTINGS_AK_ORDER_FULL_NAME, account_key_order,
+				sizeof(account_key_order));
+	if (err) {
+		LOG_ERR("Unable to save new Account Key order in Settings."
+			"Not propagating the error and keeping correct Account Key order in RAM."
+			"Account Key order may change at reboot.");
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:417]: Yes, but if more than one AK order update fails, the order may change. We will update order on commit to include all Keys in order list but the order will be more or less "random".

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:219]: Break does not cost much and it allows application to get informed about an error via the settings_load return value. I think it's worth it :)

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 7]
```diff
@@ -118,31 +141,105 @@ static bool zero_id_check(uint8_t start_idx)
...
+							     (id - ACCOUNT_KEY_CNT);
+}
+
+uint8_t next_account_key_id(void)
+{
+	if (account_key_count < ACCOUNT_KEY_CNT) {
+		return ACCOUNT_KEY_MIN_ID + account_key_count;
+	}
+
+	return toggle_ak_id(account_key_order[ACCOUNT_KEY_CNT - 1]);
```
[the remark this answers is not in the document]
Comment (alstrzebonski) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:169]: maybe `get_least_recent_key_id` :) ?

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:417]: So maybe `After a settings error the Account Key order may change at reboot` :) ?

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:169]: Sounds good

**Review by mkapala-nordic — COMMENTED** [review]
Reviewed, nothing caught my eye beside what Marek already pointed out.

**Comment by alstrzebonski** [comment]
Addressed review comments. Please review.

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 8]
```diff
@@ -153,55 +241,38 @@ static int fp_settings_commit(void)
...
+		if (account_key_id_to_idx(id) != i) {
+			return -EINVAL;
 		}
 	}
 
 	if (first_zero_idx != -1) {
-		if (!zero_id_check(first_zero_idx + 1)) {
-			return -EINVAL;
-		}
+		for (size_t i = 0; i < ACCOUNT_KEY_CNT; i++) {
```
Comment (mkapala-nordic) [inline subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:257]: Consider splitting this loop into two separate ones with different range (`<0; first_zero_idx)` and `<first_zero_idx; ACCOUNT_KEY_CNT>`). I think it could increase readability.

**Review by mkapala-nordic — COMMENTED** [review]
One minor comment, beside that lgtm.
Will anything more be done in  the scope of this PR?

**Comment by alstrzebonski** [comment]
> One minor comment, besides that lgtm. Will anything more be done in the scope of this PR?

Yes, I need to align Fast Pair tests.

**Comment by alstrzebonski** [comment]
Everything aligned. Please review.

**Comment by alstrzebonski** [comment]
Rebased to retrigger CI.

**Review by mkapala-nordic — APPROVED** [review]

**Review by MarekPieta — CHANGES_REQUESTED** [review]
One last issue and a few nits

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 9]
```diff
@@ -601,6 +601,8 @@ Bluetooth libraries and services
     * The :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_EXT_PN` Kconfig option to the :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_PN` Kconfig option.
     * The :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_EXT_PN_LEN_MAX` Kconfig option to the :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_PN_LEN_MAX` Kconfig option.
 
+  * Changed the Fast Pair storage module to overwrite the least recently used Account Key instead of the oldest Account Key on Account Key write.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:604]:
```suggestion
  * Updated the Fast Pair storage module to overwrite the least recently used Account Key instead of the oldest Account Key on Account Key write.
```

**Review by divipillai — APPROVED** [review]

**Comment by alstrzebonski** [comment]
Addressed review comments.

**Comment by alstrzebonski** [comment]
Applied fix

File: subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c [hunk 10]
```diff
@@ -107,42 +129,114 @@ static int fp_settings_set(const char *name, size_t len, settings_read_cb read_c
...
+			if (ak_order_update_count >= account_key_count) {
+				__ASSERT_NO_MSG(false);
+				return -EINVAL;
+			}
+			ak_order_update_ram(id);
+			ak_order_update_count++;
+			/* Start loop again to check whether after update no existent AK has been
+			 * lost.
+			 */
+			loop_reset = true;
```
Comment (MarekPieta) [inline subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:220]: `loop_reset` will be ignored for the last key

**Review by MarekPieta — CHANGES_REQUESTED** [review]

**Comment by alstrzebonski** [comment]
Fixed.

**Review by mkapala-nordic — APPROVED** [review]

**Comment by alstrzebonski** [comment]
Fixed bug found during testing. Please rereview.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 11]
```diff
@@ -601,6 +601,8 @@ Bluetooth libraries and services
     * The :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_EXT_PN` Kconfig option to the :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_PN` Kconfig option.
     * The :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_EXT_PN_LEN_MAX` Kconfig option to the :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_PN_LEN_MAX` Kconfig option.
```
Comment (MarekPieta) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:602]: Commit message typo: `overwrited` -> `overwritten`

**Review by MarekPieta — APPROVED** [review]

**Comment by alstrzebonski** [comment]
Fixed commit message.
